Putting it together: Relative Free Energy

proteneer opened the pull request:
This is a draft PR showing how to implement an explicit solvent relative free energy calculation. Can be used on either complex or solvent with minimal difficulty. Lambda schedules need extensive tuning, but probably good enough for a first pass.

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +742,155 @@ def setup_intermediate_state(self, lamb):
+        for k, v in enumerate(self.b_to_c):
+            c_to_b[v] = k
+
+        for idx, membership in enumerate(self.c_flags):
```
setup decoupling schemes for the ligand-host interactions

proteneer commented on timemachine/potentials/nonbonded.py:
```diff
@@ -334,7 +334,11 @@ def nonbonded_v3_on_precomputed_pairs(
     offsets: P (w,) distance offsets for electrostatic and lj interactions
+=======
+    offsets: P, 2 (w,) distance offsets for electrostatic and lj interactions
+>>>>>>> b154fd1b... WIP
```
oopsie - willfix

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,390 @@
+
+
+@pytest.mark.skip(reason="This is currently too slow to run on CI")
+def test_combine_host_guest():
```
leaving this here for now for when we setup more extensive system tests later on

mcwitt replied:
Would it make sense to run this as a nightly test?

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +742,155 @@ def setup_intermediate_state(self, lamb):
+        host_system,
+        guest_system,
+        lamb,
+    ):
```
(note the same value of `lamb` should be used when calling the actual return energy functions, mainly due to the host-guest nonbonded interactions).

maxentile commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +742,182 @@ def setup_intermediate_state(self, lamb):
+        ligand_exclusions = []
+        ligand_scale_factors = []
+
+        num_ligand_atoms = self.get_num_atoms()
```
Style nit: "ligand" and "guest" both used this stretch to refer to the hybrid molecule.

(Possibly ambiguous collection of names: "guest", "ligand", "ligand_a", "ligand_b", "mol_a", "mol_b". Would there be any loss if we say "guest" throughout this function for hybrid molecule, "mol_a", "mol_b" for the endstates, and remove names "ligand", "ligand_a", "ligand_b"?)

proteneer replied:
renamed to guest, I'm going to keep `mol_a` and `mol_b` for now. I actually didn't want to initially have the `SingleTopology` class have any host-guest semantics (and be focused only on pairs of small molecules).

maxentile commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +742,182 @@ def setup_intermediate_state(self, lamb):
+        c_to_a = dict()
+        for k, v in enumerate(self.a_to_c):
+            c_to_a[v] = k
+
+        c_to_b = dict()
+        for k, v in enumerate(self.b_to_c):
+            c_to_b[v] = k
+
+        for idx, membership in enumerate(self.c_flags):
+            if membership == 0:  # core atom
+                a_idx = c_to_a[idx]
+                b_idx = c_to_b[idx]
+                plane = 0
+                offset = 0
+                # interpolate charges when in common-core
+                q = (1 - lamb) * ligand_a_q[a_idx] + lamb * ligand_b_q[b_idx]
+                sig = (1 - lamb) * ligand_a_lj[a_idx, 0] + lamb * ligand_b_lj[b_idx, 0]
+                eps = (1 - lamb) * ligand_a_lj[a_idx, 1] + lamb * ligand_b_lj[b_idx, 1]
+            elif membership == 1:  # dummy_A
+                a_idx = c_to_a[idx]
+                plane = 0
+                offset = 1
+                q = ligand_a_q[a_idx]
+                sig = ligand_a_lj[a_idx, 0]
+                eps = ligand_a_lj[a_idx, 1]
+            elif membership == 2:  # dummy_B
+                b_idx = c_to_b[idx]
+                plane = -1
+                offset = 1
+                q = ligand_b_q[b_idx]
+                sig = ligand_b_lj[b_idx, 0]
+                eps = ligand_b_lj[b_idx, 1]
+            else:
+                assert 0
+
+            ligand_lambda_plane_idxs.append(plane)
+            ligand_lambda_offset_idxs.append(offset)
+            ligand_charges.append(q)
+            ligand_sigmas.append(sig)
+            ligand_epsilons.append(eps)
```
Organization: Would it be useful to extract this stretch into a different function `self.combine_params(a_qlj, b_qlj) -> (c_qlj, c_plane_idxs, c_offset_idxs)` or something, since it contains the main "opinionated" steps (e.g. how to interpolate core q, sig, eps)? May also be good to test this separately.

(The concatenation logic that occupies the rest of the function seems fairly closed-ended.)

proteneer replied:
I think that's a good idea - @mcwitt would that be useful for you?

maxentile commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +742,182 @@ def setup_intermediate_state(self, lamb):
+            i) at lambda = 0, the dummy atoms of mol_a are fully interacting, and become non-interacting at lambda = 1
+            ii) at lambda = 0, the dummy atoms of mol_b are non-interacting, and become fully interacting at lambda = 1
+            iii) the core atoms have parameters interpolated from mol_a's qlj to mol_b's qlj.
+        3) guest-guest interactions use pre-computed nonbonded interactions implemented as a pairlist.
```
Initially I didn't follow how this function implements 3) -- does this happen elsewhere?

maxentile commented on timemachine/cpp/CMakeLists.txt:
```diff
@@ -12,14 +12,7 @@ find_package(PythonLibs 3.7 REQUIRED)
-  " -Wno-unused-parameter"
-  " -Wreorder"
-)
+
```
Wasn't mentioned in PR description -- confirm if this is part of the current PR?

proteneer replied:
good catch, this was a bad merge conflict resolution. will fix

maxentile reviewed (commented):
Main substantive aspect I didn't follow yet is where the guest-guest pairlist is constructed in `combine_with_host`, otherwise minor comments!

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +742,182 @@ def setup_intermediate_state(self, lamb):
+
+        Returns
+        -------
+        7-tuple of potentials
+            bond, angle, torsion, chiral_atom, chiral_bond, guest_nonbonded_precomputed, full_nonbonded
```
How about `HostGuestSystem`?

proteneer commented on timemachine/testsystems/relative.py:
```diff
@@ -66,4 +63,55 @@ def setup_hif2a_ligand_pair(ff="smirnoff_1_1_0_ccc.py"):
     return free_energy.RelativeFreeEnergy(single_topology)
 
 
+def get_hif2a_ligand_pair_single_topology(ff="smirnoff_1_1_0_ccc.py"):
```
yes! but I'm going to do this in a separate PR to avoid cluttering this one.

proteneer replied:
(there is a much bigger PR coming to completely remove single topology)

mcwitt commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +743,191 @@ def setup_intermediate_state(self, lamb):
+            combined_torsion_params = np.concatenate([host_torsion.params, guest_torsion.params])
+        else:
+            combined_torsion_idxs = np.array(guest_torsion.get_idxs(), dtype=np.int32) + num_host_atoms
+            combined_torsion_params = np.array(guest_torsion.params, dtype=np.float64)
```
It's not obvious to me why a special case is needed for torsions. Worth adding a comment?

proteneer replied:
solvent host doesn't have torsions, complex host does. will add comment

mcwitt commented on timemachine/fe/system.py:
```diff
@@ -149,3 +142,29 @@ def U_fn(x):
+            self.chiral_bond,
+            self.nonbonded_guest_pairs,
+            self.nonbonded_host_guest,
+        ]
```
nit: could be a dataclass? (can use `Any` for attributes that don't have well-defined types)

proteneer replied:
probably - but I think I'd rather you do this correctly (along with the `VacuumSystem` class).

mcwitt commented on timemachine/testsystems/dhfr.py:
```diff
@@ -20,3 +20,14 @@ def setup_dhfr():
+    host_system = protein_ff.createSystem(
+        host_pdb.topology, nonbondedMethod=app.NoCutoff, constraints=None, rigidWater=False
+    )
+    return host_system
```
Could `setup_dhfr` call this to avoid repetition?

proteneer replied:
yes, willfix - got a bit too zealous on the copy and pasting

proteneer replied:
ugh - this is not quite straightforward, because the `setup` method references things like `host_pdb`, etc. which are not in the `host_system`

I don't want to make this function too complicated, so i'm going to settle for duplication of code here.

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+        elif isinstance(pot, potentials.Nonbonded):
+            host_system.nonbonded = pot
+        else:
+            assert 0, "Unknown potential"
```
yes - this whole function can be removed actually

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+
+    temperature = 300.0
+    friction = 1.0
+    masses = np.ones(total_num_atoms) * 5  # fixme
```
oops - forgot to fix this!

proteneer replied:
ah I remember why I did it this way. I'm not sure how to set the masses here for core atoms. We have several choices:

1) For core atoms, at lambda = 0, use mol_a's masses. at lambda=1 use mol_b's masses. At 0<lambda<1 linearly interpolate
2) Use only mol_a or only mol_b 
3) Optimize masses at each lambda window based on the interpolated bond and angle parameters.

@maxentile @jkausrelay thoughts?

proteneer replied:
in light of not wanting to hold this up, I'll implement 1) now and re-visit this later

proteneer replied:
I'm going to implement 4) take the max

maxentile replied:
4) sounds safer than 1-3) for now, to avoid introducing any complications in HREX etc. from having different masses in different lambda windows.

In a later pass, 3) might buy some efficiency improvement

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+def _test_combined_parameters_impl(host_system_omm):
+    # test bonded and nonbonded parameters are correct at the end-states.
+    # 1) we expected bonded idxs in the ligand to be shifted by num_host_atoms
+    # 2) we expected nonbonded lambda_idxs to be shifted
```
willfix

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+
+    for lamb in [0.0, 1.0]:
+
+        hgs = st3.combine_with_host(host_sys, lamb=lamb)
```
willfix

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+                if i < j:
+                    expected_guest_idxs.add((i + num_host_atoms, j + num_host_atoms))
+
+        assert hgs_exc_guest_idxs == hgs_exc_guest_idxs
```
good call - linter missed this

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+
+        assert hgs_exc_guest_idxs == hgs_exc_guest_idxs
+
+        # 2) decoupling parameters for host-guest interactions
```
maybe we can break up the bonded vs nonbonded bits?

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+        mol_b_charges = st3.ff.q_handle.parameterize(st3.mol_b)
+        mol_b_sig_eps = st3.ff.lj_handle.parameterize(st3.mol_b)
+
+        c_to_a = dict()
+        for k, v in enumerate(st3.a_to_c):
+            c_to_a[v] = k
```
oh this is nice! I actually movedthis in the constructor but I'll adapt this new style. I wasn't sure how to do the initialization this way!

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -735,3 +743,191 @@ def setup_intermediate_state(self, lamb):
+        guest_torsion = guest_system.torsion
+        guest_nonbonded = guest_system.nonbonded
+        guest_chiral_atom = guest_system.chiral_atom
+        guest_chiral_bond = guest_system.chiral_bond
```
sure

proteneer commented on tests/test_single_topology_combined.py:
```diff
@@ -0,0 +1,381 @@
+# test code for combining host guest systems
+import matplotlib.pyplot as plt
```
going to move the simulation test code into a scripts folder - sorry this was sloppy on my end!

proteneer commented:
I've moved the expensive test out of the tests folder and refactored into a script in `examples/relative_free_energy.py` using the `hif2a testsystem`. This also supports complex calculations.
